File: raytrace/Object/Shape/Sphere.cpp
#include "Object/Shape/Sphere.hpp"
#include "Math/Transformation.hpp"

#include <math.h>

namespace Object {
namespace Shape {

std::unique_ptr<Sphere> Sphere::fromAst(AST *ast)
{
	Math::Point position(ast->children[0]->children[0]->data._vector);
	float radius = ast->children[0]->children[1]->data._float;
	std::unique_ptr<Sphere> sphere = std::make_unique<Sphere>(position, radius);

	return sphere;
}

Sphere::Sphere(const Math::Point &position, float radius)
	: mPosition(position), mRadius(radius)
{
}

bool Sphere::intersect(const Math::Ray &ray, float &distance, Math::Normal &normal) const
{
	float a, b, c;
	float disc;

	a = ray.direction().magnitude2();
	b = 2 * ((ray.origin() - mPosition) * ray.direction());
	c = (ray.origin() - mPosition).magnitude2() - mRadius * mRadius;

	disc = b * b - 4 * a * c;
	if(disc >= 0)
	{
		float newDistance = (-b - sqrt(disc)) / (2 * a);

		if(newDistance >= 0 && newDistance < distance)
		{
			distance = newDistance;
			Math::Point point = ray.origin() + ray.direction() * distance;
			normal = Math::Normal(point - mPosition) / mRadius;
			return true;
		}

		newDistance = (-b + sqrt(disc)) / (2 * a);

		if(newDistance >= 0 && newDistance < distance)
		{
			distance = newDistance;
			Math::Point point = ray.origin() + ray.direction() * distance;
			normal = Math::Normal(point - mPosition) / mRadius;
			return true;
		}
	}

	return false;
}

BoundingVolume Sphere::boundingVolume(const Math::Transformation &transformation) const
{
	float mins[BoundingVolume::NUM_VECTORS];
	float maxes[BoundingVolume::NUM_VECTORS];

	for(int i=0; i<BoundingVolume::NUM_VECTORS; i++) {
		const Math::Vector &vector = BoundingVolume::vectors()[i];
		float x = Math::Vector(mPosition) * vector;
		float y = mRadius * vector.magnitude();

		mins[i] = x - y;
		maxes[i] = x + y;
	}

	return BoundingVolume(mins, maxes);
}

bool Sphere::sample(float u, float v, Math::Point &point, Math::Vector &du, Math::Vector &dv, Math::Normal &normal) const
{
	return false;
}

}
}

File: src/object/shape/sphere.rs
use crate::geo;
use crate::object;

use geo::Bivec3;
use geo::Normal3;
use geo::Point2;
use geo::Point3;
use geo::Ray;
use geo::Transformation;
use geo::Vec3;

use object::BoundingVolume;
use object::Shape;
use object::ShapeIntersection;

pub struct Sphere {
    position: Point3,
    radius: f32,
}

impl Sphere {
    pub fn new(position: Point3, radius: f32) -> Sphere {
        Sphere { position, radius }
    }
}

impl Shape for Sphere {
    fn intersect(&self, ray: Ray, max_distance: f32, _closest: bool) -> Option<ShapeIntersection> {
        let a = ray.direction.mag2();
        let b = 2.0 * (ray.origin - self.position) * ray.direction;
        let c = (ray.origin - self.position).mag2() - self.radius * self.radius;
        let disc = b * b - 4.0 * a * c;

        if disc >= 0.0 {
            let dist0 = (-b - f32::sqrt(disc)) / (2.0 * a);
            let dist1 = (-b + f32::sqrt(disc)) / (2.0 * a);
            for distance in [dist0, dist1] {
                if distance >= 0.0 && distance < max_distance {
                    let point = ray.origin + ray.direction * distance;
                    let normal = Normal3::from(point - self.position) / self.radius;

                    return Some(ShapeIntersection::new(
                        distance,
                        normal,
                        Bivec3::ZERO,
                        Point2::ZERO,
                    ));
                }
            }
        }

        return None;
    }

    fn bounding_volume(&self, xform: Transformation) -> BoundingVolume {
        let mut mins = [0.0; BoundingVolume::NUM_VECTORS];
        let mut maxes = [0.0; BoundingVolume::NUM_VECTORS];

        for i in 0..BoundingVolume::NUM_VECTORS {
            let vector = BoundingVolume::VECTORS[i];
            let x = Vec3::from(self.position) * vector;
            let y = self.radius * vector.mag();

            mins[i] = x - y;
            maxes[i] = x + y;
        }

        return BoundingVolume::with_mins_maxes(mins, maxes);
    }
}
